File: go/jzoffer/exchange_21.go
package jzoffer

func exchange(nums []int) []int {
	if nums == nil || len(nums) == 0 {
		return []int{}
	}

	i, j := 0, len(nums)-1

	for i < j {
		// 前面是奇数就跳过
		if nums[i]&1 == 1 {
			i++
			continue
		}
		if nums[j]&1 == 0 {
			j--
			continue
		}
		swap(nums, i, j)
	}

	return nums
}

File: java/jzoffer/chapter3/Exchange_21.java
package chapter3;

import java.util.Arrays;

public class Exchange_21 {

    public int[] exchange(int[] nums) {
        int i = 0, j = nums.length - 1;

        while (i < j) {
            while (i < j && (nums[i] & 1) == 1) {
                i++;
            }
            while (i < j && (nums[j] & 1) == 0) {
                j--;
            }
            if (i >= j) break;

            if (nums[i] != nums[j]) {
                nums[i] = nums[i] ^ nums[j];
                nums[j] = nums[i] ^ nums[j];
                nums[i] = nums[i] ^ nums[j];
            }
            i++;
            j--;
        }
        return nums;
    }

    public static void main(String[] args) {
        System.out.println(Arrays.toString(new Exchange_21().exchange(new int[]{1, 3, 5})));
    }
}
